File: sliding_window/1004_max_consecutive_ones_3.go
func longestOnes(nums []int, k int) int {
	begin := 0
	windowState := 0 // количество встреченых нулей
	result := 0

	for end := range len(nums) {
		if nums[end] == 0 {
			windowState++
		}

		// если был хотя бы один ноль то отсуюад выйдем
		for windowState > k {
			if nums[begin] == 0 {
				windowState--
			}
			begin++
		}

		result = max(result, end-begin+1)
	}

	return result
}

File: sliding_window/1004_max_consecutive_ones_3.py
def longestOnes(nums: list[int], k: int) -> int:
	begin = 0
	window_state = 0
	result = 0

	for end, n in enumerate(nums):
		if n == 0:
			window_state += 1
		while window_state > k:
			if nums[begin] == 0:
				window_state -= 1
			begin += 1

		result = max(result, end - begin + 1)
	return result
